core/src/main/java/org/infinispan/container/AbstractSegmentedDataContainer.java

```diff
@@ -0,0 +1,634 @@
+      for (PrimitiveIterator.OfInt iter = segments.iterator(); iter.hasNext(); ) {
+         int segment = iter.next();
+         ConcurrentMap<K, InternalCacheEntry<K, V>> map = getMapForSegment(segment);
+         size += map.size();
+         if (size < 0) return Integer.MAX_VALUE;
+      }
+      return size;
+   }
+
+   @Override
+   public void evict(int segment, K key) {
+      ConcurrentMap<K, InternalCacheEntry<K, V>> entries = getMapForSegment(segment);
+      if (entries != null) {
+         entries.computeIfPresent(key, (o, entry) -> {
+            passivator.passivate(entry);
+            computeEntryRemoved(o, entry);
+            return null;
+         });
+      }
+   }
+
+   @Override
+   public void evict(K key) {
+      evict(getSegmentForKey(key), key);
+   }
+
+   @Override
+   public InternalCacheEntry<K, V> compute(int segment, K key, ComputeAction<K, V> action) {
+      ConcurrentMap<K, InternalCacheEntry<K, V>> entries = getMapForSegment(segment);
+      return entries != null ? entries.compute(key, (k, oldEntry) -> {
+         InternalCacheEntry<K, V> newEntry = action.compute(k, oldEntry, entryFactory);
+         if (newEntry == oldEntry) {
+            return oldEntry;
+         } else if (newEntry == null) {
+            computeEntryRemoved(k, oldEntry);
+            activator.onRemove(k, false);
+            return null;
+         }
+         computeEntryWritten(k, newEntry);
+         activator.onUpdate(k, oldEntry == null);
+         if (trace())
+            log().tracef("Store %s in container", newEntry);
+         return newEntry;
+      }) : null;
+   }
+
+   @Override
+   public InternalCacheEntry<K, V> compute(K key, ComputeAction<K, V> action) {
+      return compute(getSegmentForKey(key), key, action);
+   }
+
+   @Override
+   public Spliterator<InternalCacheEntry<K, V>> spliterator(IntSet segments) {
+      return Spliterators.spliteratorUnknownSize(iterator(segments), Spliterator.DISTINCT | Spliterator.CONCURRENT | Spliterator.NONNULL);
+   }
+
+   @Override
+   public Spliterator<InternalCacheEntry<K, V>> spliterator() {
+      return Spliterators.spliteratorUnknownSize(iterator(), Spliterator.DISTINCT | Spliterator.CONCURRENT | Spliterator.NONNULL);
+   }
+
+   @Override
+   public Spliterator<InternalCacheEntry<K, V>> spliteratorIncludingExpired(IntSet segments) {
+      return Spliterators.spliteratorUnknownSize(iteratorIncludingExpired(segments), Spliterator.DISTINCT | Spliterator.CONCURRENT | Spliterator.NONNULL);
+   }
+
+   @Override
+   public Spliterator<InternalCacheEntry<K, V>> spliteratorIncludingExpired() {
+      return Spliterators.spliteratorUnknownSize(iteratorIncludingExpired(), Spliterator.DISTINCT | Spliterator.CONCURRENT | Spliterator.NONNULL);
+   }
+
+   @Override
+   public void clear(IntSet segments) {
+      Iterator<InternalCacheEntry<K, V>> iter = iteratorIncludingExpired(segments);
+      while (iter.hasNext()) {
+         iter.next();
+         iter.remove();
```

Comment:
Hrmm that should work. Guessing I just didn't think about it too hard :)

Reply:
Ah it was because of `DefaultDataContainer` - I will change this and try to have `DefaultDataContainer` no longer implement `SegmentedDataContainer` afterwards as mentioned above.